server/src/test/java/org/apache/druid/segment/indexing/DataSchemaTest.java

```diff
@@ -324,6 +324,109 @@ public void testEmptyDatasource()
     );
   }
 
+  @Test
+  public void testNewlineDatasource()
+  {
+    Map<String, Object> parser = jsonMapper.convertValue(
+        new StringInputRowParser(
+            new JSONParseSpec(
+                new TimestampSpec("time", "auto", null),
+                new DimensionsSpec(
+                    DimensionsSpec.getDefaultSchemas(ImmutableList.of("time", "dimA", "dimB", "col2")),
+                    ImmutableList.of("dimC"),
+                    null
+                ),
+                null,
+                null
+            ),
+            null
+        ), JacksonUtils.TYPE_REFERENCE_MAP_STRING_OBJECT
+    );
+
+    expectedException.expect(IllegalArgumentException.class);
+    expectedException.expectMessage(
+        "dataSource cannot contain whitespace character."
+    );
+
+    DataSchema schema = new DataSchema(
+        "new\nline\ncharacter",
+        parser,
+        null,
+        null,
+        null,
+        jsonMapper
+    );
+  }
+
+
+  @Test
+  public void testCarriageReturnDatasource()
```

Comment:
Since the logic is very similar for the new tests you added, consider refactoring to have a private helper method.